CONVENTIONS.md

```diff
@@ -126,6 +126,22 @@ it is used to manage all elements of the cluster.
 * Every component must remain available to consumers without disruption during upgrades and reconfiguration
   * See the next section for more details
 
+#### High Availability
+
+We focus on minimizing the impact of a failure of individual nodes in a cluster by ensuring operators or operands are spread across multiple nodes while maintaining backwards compatibility.
+The smallest supported configuration for OpenShift is 3 control-plane nodes and 2 workers, or 3 control-plane nodes that also run workloads (compact clusters), so the following scenarios are intended to support the minimum 2 worker node configuration
+
+
+* If the operator or operand wishes to be highly available and can tolerate the loss of one replica, the default configuration for it should be
+  * Two replicas
+  * Hard pod [anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#always-co-located-in-the-same-node) on hostname (no two pods on the same node)
+  * Use the maxUnavailable rollout strategy on deployments (prefer 25% by default for a value)
+  * This is the recommended approach for all the components unless the component needs 3 replicas
+* If the operator or operand requires >= 3 replicas and should be running on worker nodes
+  * Set soft pod anti-affinity on the hostname so that pods prefer to land on separate nodes (will be violated on two node clusters)
+  * Use maxSurge for deployments if possible since the spreading rules are soft
```

Comment:
Why use maxSurge important when using soft affinity?  Is there a reason not to use maxSurge with hard affinity?

Reply:
maxSurge with hard affinity would just allow pods to be created but because of node limit the pods will be in pending state not consuming actual resources on the node whereas with soft affinity since the constraints are ok to be violated pods will be created on the nodes after scheduling stage

Reply:
Ah, I see.  Thanks for explaining!  The way we get around that with the ingress operator is that it creates a deployment with a label that has a hash of the deployment and a hard pod anti-affinity rule that selects on that label, so that the hard anti-affinity rule does not apply to changes to the deployment that cause rolling updates.  Maybe it isn't worth going into all that detail here though.